feat: use build images to run the application

Yoan-MA opened the pull request:
Instead of building images on the fly, we'll now pulling images from a Container Registry

Yoan-MA committed:
feat: use build images to run the application

Instead of building images on the fly, we'll now pulling
images from a Container Registry

Yoan-MA committed:
feat: add github action workflow to push new images

When there is a new push on Master, the Github Action workflow will be triggered
to build images, tag them as latest and push them to the container Registry

Yoan-MA committed:
feat: use github container registry instead of docker hub

Yoan-MA committed:
chore: activate workflow for a pr to test it

fix: syntax error

fix the following error
The workflow is not valid. .github/workflows/deploy-image-on-master-merge.yaml (Line: 25, Col: 20):
Unrecognized named-value: 'env'. Located at position 1 within expression: env.REGISTRY

chore: lowercase repository name

chore: fix syntax error

Related error
The workflow is not valid. .github/workflows/deploy-image-on-master-merge.yaml (Line: 72, Col: 17):
Unrecognized named-value: 'REPOSITORY_NAME_PREFIX_LOWER_CASE'. Located at position 1 within
expression: REPOSITORY_NAME_PREFIX_LOWER_CASE

Yoan-MA committed:
fix: Change built platform

corenting commented on .github/workflows/deploy-image-on-master-merge.yaml:
```diff
@@ -0,0 +1,79 @@
+          platforms: linux/amd64, linux/arm/v7, linux/arm64/v8
+          file: ${{ matrix.dockerfile }}
+          push: true
+          tags: |
```
We could use https://github.com/docker/metadata-action to generate the tags beforehand in the workflow.

corenting commented on .github/workflows/deploy-image-on-master-merge.yaml:
```diff
@@ -0,0 +1,79 @@
+        id: docker_build
+        uses: docker/build-push-action@v2
+        with:
+          platforms: linux/amd64, linux/arm/v7, linux/arm64/v8
```
I think we can limit ourselves to linux/amd64 and linux/arm64/v8

corenting commented on .github/workflows/deploy-image-on-master-merge.yaml:
```diff
@@ -0,0 +1,79 @@
+      # Use Emulator to build images on different architectures
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@master
+        with:
+          platforms: all
```
I don't think we need to specify the argument, it is the default value

Yoan-MA committed:
chore: take care about PR feedbacks

Yoan-MA committed:
chore: fix error when building the listinapi image

Define a context to find files
Github Action crashes because of
Error: buildx failed with: ERROR: failed to solve: failed to compute cache key: failed to
calculate checksum of ref 6siri1cgvcqkb84vde7l39f20::2ykz5f6l7ay6eyd6fxzxy1uxe: "/Pipfile.lock": not found

feat: define hardcoded user:group for pricemap-python

chore: define default user:group on each containers

Yoan-MA committed:
chore: change container registry

NB : For testing purpose, we're using tag pr-22 instead of latest

Yoan-MA committed:
chore: add tag latest when merged on master

Yoan-MA committed:
chore: clean up branch

Use latest tag

quentinus95 commented on .github/workflows/deploy-image-on-master-merge.yaml:
```diff
@@ -0,0 +1,82 @@
+        uses: docker/metadata-action@v4
+        with:
+          images: |
+            ${{ env.REGISTRY }}/${{ steps.lowercase.outputs.repository_name_lower_case }}/${{ matrix.image }}
```
@Yoan-MA https://github.com/docker/build-push-action/issues/237#issuecomment-848673650

https://github.com/docker/metadata-action#image-name-and-tag-sanitization

> To ease the integration in your workflow, this action will automatically:
> 
> Lowercase the image name

quentinus95 commented on docker-compose.yml:
```diff
@@ -2,7 +2,8 @@ version: "3.4"
   listingapi:
-    build: listingapi
+    image: ghcr.io/meilleursagents/backend-technical-test/listingapi:latest
+    user: "listingapi:group_listingapi"
```
Most of the time it is conventional to have the group matching the user, do you see situations where having the `group_` prefix can be useful?

Yoan-MA replied:
Nope.

Yoan-MA committed:
chore: change group name into containers

fix: add user to group

Yoan-MA committed:
chore: improve github action

Remove lowercase step (already included by the metadata step)
see https://github.com/docker/metadata-action#image-name-and-tag-sanitization
